### PR TITLE
Make app2app spec clearer & support migration to app2app from existing sessions

### PR DESCRIPTION
**Pull request by tung2744**

### CODE REVIEW

**Comment by tung2744** [comment]
@louischan-oursky Updated the spec for the recently discussed config for migration of existing session.

File: docs/specs/app2app.md [hunk 1]
```diff
@@ -21,20 +21,38 @@ An app can start the authentication flow by opening a link to another app, inste
...
+  - A app2app section will be added to the client configs.
+
+  ```yaml
+  oauth:
+    clients:
+      - client_id: CLIENT_ID
+        # ...Other configs
+        x_app2app:
+          enabled: true
+          allow_insecure_device_key_binding: false
```
Comment (louischan-oursky) [inline docs/specs/app2app.md:35]: The convention in client metadata is not to use nested object. So I think we should call them `x_app2app_enabled` and `x_app2app_insecure_device_key_binding_enabled`.

File: docs/specs/app2app.md [hunk 2]
```diff
@@ -21,20 +21,38 @@ An app can start the authentication flow by opening a link to another app, inste
...
+  ```
+
+  - `enabled`: boolean. Whether the client is able to handle app2app authentication requests from other clients.
+  - `allow_insecure_device_key_binding`: boolean. Default `false`. If `true`, refresh tokens of this client without a binded `device_key` can be binded to a new `device_key` during refreshing the tokens. This option is for allowing existing logged in users to use app2app without requiring them to re-login. However this also allows sessions without a binded `device_key` to be binded with any key and participate in app2app authentication which might be a security concern.
```
Comment (louischan-oursky) [inline docs/specs/app2app.md:39]: s/binded/bound

File: docs/specs/app2app.md [hunk 3]
```diff
@@ -21,20 +21,38 @@ An app can start the authentication flow by opening a link to another app, inste
...
 
 - Token Endpoint
 
+  - For `grant_type=authorization_code`, a new parameter `device_key` is supported. This value can be specified by the client to enable app2app login. If specified, it should be a base64 encoded jwk of public key of a key pair, which the private key should be stored in a secure storage in the device. Such as the Android keystore or the iOS keychain. `device_key` will be ignored if `x_app2app.enabled` of the client configuration is not `true`.
```
Comment (louischan-oursky) [inline docs/specs/app2app.md:43]: Rename this to `x_app2app_device_key_jwt` to signify it is our own extension.

And we are going to make it a JWT.

header
```
{
  "alg": "ES256",
  "kty": "EC",
  "kid": "randomly-generated-kid",
  "jwk": { ... }
}
```

Payload
```
{
  "iat": 123,
  "exp": 123,
   "challenge": "one-time-use-challenge-obtain-from-server"
}
```

File: docs/specs/app2app.md [hunk 4]
```diff
@@ -21,20 +21,38 @@ An app can start the authentication flow by opening a link to another app, inste
...
+
+    - Multiple key types such as RS256 and NIST P-256 are supported.
+
+  - For `grant_type=refresh_token`, a new parameter `device_key` is supported if `x_app2app.allow_insecure_device_key_binding` is `true`. The provided `device_key` will be binded to the provided refresh token, if the refresh token isn't already binded to a `device_key`. If the refresh token was already binded to a `device_key` which doesn't match the one provided in this parameter, an error will be returned.
```
Comment (louischan-oursky) [inline docs/specs/app2app.md:47]: Apply the same change.

File: docs/specs/app2app.md (hunk 3 above)
Comment (tung2744) [reply docs/specs/app2app.md:43]: @louischan-oursky Why making it a jwt? This is the public key to bind to the refresh token but not to verify the user owns the private key.

Comment (louischan-oursky) [reply docs/specs/app2app.md:43]: 1. It is a proof that the caller has the private key. We will verify the signature of the JWT with the JWK inside it.
2. Avoid replay attack. We will check if the challenge is consumed or not.

**Comment by tung2744** [comment]
@louischan-oursky Updated, thanks :adore:

**Pull request merged**